src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs

```diff
@@ -56,20 +57,27 @@ public class VkontakteAuthenticationHandler : OAuthHandler<VkontakteAuthenticati
                 throw new HttpRequestException("An error occurred while retrieving the user profile.");
             }
 
-            var container = JObject.Parse(await response.Content.ReadAsStringAsync());
-            var payload = container["response"].First as JObject;
-
-            if (tokens.Response["email"] != null)
+            using (var container = JsonDocument.Parse(await response.Content.ReadAsStringAsync()))
             {
-                payload.Add("email", tokens.Response["email"]);
-            }
+                JsonElement payload;
+
+                using (var enumerator = container.RootElement.GetProperty("response").EnumerateArray())
+                {
+                    payload = enumerator.First();
+                }
 
-            var principal = new ClaimsPrincipal(identity);
-            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
-            context.RunClaimActions(payload);
+                var principal = new ClaimsPrincipal(identity);
+                var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
+                context.RunClaimActions();
 
-            await Options.Events.CreatingTicket(context);
-            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+                if (tokens.Response.RootElement.TryGetProperty("email", out var email))
+                {
+                    identity.AddClaim(new Claim(ClaimTypes.Email, email.GetString(), ClaimValueTypes.String, ClaimsIssuer));
```

Comment:
How about calling RunClaimActions again and pass in `tokens.Response.RootElement`?